src/scheduler/DistributedCoordination.cpp

```diff
@@ -0,0 +1,236 @@
+#include <faabric/scheduler/DistributedCoordination.h>
+#include <faabric/util/timing.h>
+
+#define GROUP_TIMEOUT_MS 20000
+
+#define DISTRIBUTED_SYNC_OP(opLocal, opRemote)                                 \
+    {                                                                          \
+        int32_t appId = msg.appid();                                           \
+        if (msg.masterhost() == sch.getThisHost()) {                           \
+            opLocal(appId);                                                    \
+        } else {                                                               \
+            std::string host = msg.masterhost();                               \
+            faabric::scheduler::FunctionCallClient& client =                   \
+              sch.getFunctionCallClient(host);                                 \
+            opRemote(appId);                                                   \
+        }                                                                      \
+    }
+
+#define FROM_MAP(varName, T, m, ...)                                           \
+    {                                                                          \
+        if (m.find(appId) == m.end()) {                                        \
+            faabric::util::FullLock lock(sharedMutex);                         \
+            if (m.find(appId) == m.end()) {                                    \
+                m[appId] = std::make_shared<T>(__VA_ARGS__);                   \
+            }                                                                  \
+        }                                                                      \
+    }                                                                          \
+    std::shared_ptr<T> varName;                                                \
+    {                                                                          \
+        faabric::util::SharedLock lock(sharedMutex);                           \
+        varName = m[appId];                                                    \
+    }
+
+namespace faabric::scheduler {
+
+DistributedCoordination& getDistributedCoordination()
+{
+    static DistributedCoordination sync;
+    return sync;
+}
+
+DistributedCoordination::DistributedCoordination()
+  : sch(faabric::scheduler::getScheduler())
+{}
+
+void DistributedCoordination::setAppSize(const faabric::Message& msg,
+                                         int appSize)
+{
+    if (msg.masterhost() != sch.getThisHost()) {
+        SPDLOG_ERROR("Setting group {} size not on master ({} != {})",
+                     msg.appid(),
+                     msg.masterhost(),
+                     sch.getThisHost());
+
+        throw std::runtime_error("Setting sync group size on non-master");
+    }
+
+    appSizes[msg.appid()] = appSize;
```

Comment:
Unfortunately adding to a map isn't atomic (I guess it needs to do the copy and update the index). We've seen issues in the past where you get either a partially initialised object or a segfault under heavy contention.

Reply:
Sorry, there was a typo in the comment. I meant the map **can't** be concurrently modified.

It looks to me that we don't protect against concurrent modifications here.

Reply:
Haha ah yes, I was looking at this in isolation 😄. In its current state this is ok, as this would only be called once, by a function before spawning a group of chained functions that would do the coordination. However, this bit of code is getting removed in my current rework, so no need to worry.